### ISSUE TITLE
Cannot provide multidimensional array as argument with @MethodSource

### ISSUE THREAD

**Issue by thauer** [junit-team/junit5#1665]
## Overview

I am attempting to pass two-dimensional arrays of primitives to a parameterized test. The one-dimensional array works as expected, but JUnit complains on the second with "Error resolving parameter at index 0". This is on JUnit Jupiter 5.1.0.

[According to Sormuras](https://stackoverflow.com/questions/53133979/can-multidimensional-arrays-be-passed-via-methodsource-junit-5-1), on `5.4.0-SNAPSHOT`, the error message reads: `org.junit.jupiter.api.extension.ParameterResolutionException: Error converting parameter at index 0: No implicit conversion to convert object of type [I to type [[I` 

see the snippet below

## Steps to reproduce

```java
// This is ok
static Stream<int[]> arrayStream1(){
    return Stream.of( new int[] {1}, new int[] {2});
}

@ParameterizedTest
@MethodSource("arrayStream1")
void test1(int[] par) {
    assertTrue(true);
}

// This is not
static Stream<int[][]> arrayStream2(){
    return Stream.of( new int[][] {{1,2}}, new int[][] {{2,3}});
}

@ParameterizedTest
@MethodSource("arrayStream2")
void test2(int[][] par) {
    assertTrue(true);
}   
```

## Context

 - Used versions (Jupiter/Vintage/Platform): JUnit 5.1.0
 - Build Tool/IDE: Eclipse 4.9.0

## Deliverables

- [ ] Ensure that streams of multidimensional arrays (e.g., `Stream<int[][]>` may be returned from `@MethodSource` factory methods.
- [ ] Document in Release Notes

**Comment by sormuras** [comment]
There exists an integration test for 2D arrays, but it is targetting a test template with two parameters:

https://github.com/junit-team/junit5/blob/master/junit-jupiter-params/src/test/java/org/junit/jupiter/params/ParameterizedTestIntegrationTests.java#L319-L327

**Comment by sormuras** [comment]
Using an `ArgumentsAccessor` as the parameter...
```java
	@ParameterizedTest
	@MethodSource("arrayStream2") void test2(ArgumentsAccessor accessor) {
		for (var object : accessor.toList()) {
			System.out.print(object.getClass() + " -> " + object);
			if (object instanceof int[]) {
				System.out.println(" = " + Arrays.toString((int[]) object));
			}
			else {
				System.out.println(" ?!");
			}
		}
	}
```
....yields:
```
class [I -> [I@8e0379d = [1, 2]
class [I -> [I@161b062a = [2, 3]
```

**Comment by sbrannen** [comment]
We currently do not support multi-dimensional arrays within a `Stream` returned by a factory method used with `@MethodSource`.

The workaround is to generate a stream of arguments (i.e., `Stream<Arguments>`) as follows.

```java
	static Stream<Arguments> arrayStream2() {
		return Stream.of(arguments((Object) new Object[][] { { 1, 2 } }),
			arguments((Object) new Object[][] { { 2, 3 } }));
	}

	@ParameterizedTest
	@MethodSource("arrayStream2")
	void test2(Object[][] par) {
		assertTrue(true);
	}
```

**Comment by sbrannen** [comment]
In case you want to see that... it happens here:

https://github.com/junit-team/junit5/blob/685c14dffa50f04e304cc305ad5ee9a5a36a33ba/junit-jupiter-params/src/main/java/org/junit/jupiter/params/provider/MethodArgumentsProvider.java#L86-L94

FYI: It's the final return statement.

**Comment by sbrannen** [comment]
_in progress_: I'm investigating a fix.

**Comment by sbrannen** [comment]
And the following is a revised version of the `Arguments` approach that retains the desired `int[][]` parameter type.

```java
static Stream<Arguments> arrayStream4() {
	return Stream.of(arguments((Object) new int[][] { { 1, 2 } }), arguments((Object) new int[][] { { 2, 3 } }));
}

@ParameterizedTest
@MethodSource("arrayStream4")
void test4(int[][] par) {
	System.err.println(Arrays.deepToString(par));
}
```

**Comment by sbrannen** [comment]
FYI: I've pushed a _WIP_ to the following feature branch.

https://github.com/junit-team/junit5/commits/issues/1665-method-source-n-dimensional-arrays

@junit-team/junit-lambda, now up for _team discussion_.

**Comment by sbrannen** [comment]
FWIW, my _fix_ on the feature branch does in fact support the failing use case described in this issue.

However, it also brings to light some oddities with regard to support for Object and primitive arrays.

1. If you have a factory method that returns `Stream<Object[]>`, that will **NOT** allow you to have a `@ParameterizedTest` method that accepts a single `Object[]` argument.
  - Rather, if the length of the array is 2, that would require a `@ParameterizedTest` method with a signature of `void test(Object, Object)`.
2. If you have a factory method that returns `Stream<int[]>`, that **WILL** allow you to have a `@ParameterizedTest` method that accepts a single `int[]` argument.
  - And that was the case before my _fix_.

**Comment by sbrannen** [comment]
> I am attempting to pass two-dimensional arrays of primitives to a parameterized test. The one-dimensional array works as expected,

@junit-team/junit-lambda, this is one of the points we need to decide on: what is actually "expected" here with regard to the previous points I made.

**Comment by sbrannen** [comment]
This has been fixed in `master` in commit c815758a57099ec646b5e60b7d12f94190e4019a.
